### DISCUSSION TITLE
URL fragment that are used to generate "Text fragments" are being rewrite by Nextjs

### DISCUSSION THREAD

**Discussion by kouliavtsev** [vercel/next.js#71970] [Help]
### Summary

https://developer.mozilla.org/en-US/docs/Web/URI/Fragment/Text_fragments

When you paste these URLs into the browser:

- [https://example.com#:~:text=for](https://example.com/#:~:text=for) scrolls to and highlights the first instance of the text for in the document.
- https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#:~:text=human scrolls to and highlights the first instance of the text human in the document.
- https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#:~:text=linked%20URL scrolls to and highlights the first instance of the text linked URL in the document.

Nextjs rewrite these URLs as:

- https://example.com
- https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a
- https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a

Is there a way to retain the original URL?

**Comment by aniruddhaadak80** [comment]
Hi @kouliavtsev ,

I understand that when using URLs with text fragments, Next.js rewrites them, resulting in the loss of the fragment information. Unfortunately, this behavior is not currently supported in Next.js.

To retain the original URLs with text fragments, consider the following approaches:

1. **Custom Server Logic**: Implement custom logic to handle requests and retain the fragment.
2. **Redirects**: Use redirects to point users to the correct URL with the fragment.
3. **Client-Side Handling**: Use client-side JavaScript to capture and process the URL fragment.


Thanks .

**Comment by kouliavtsev** [comment]
@aniruddhaadak80 

> To retain the original URLs with text fragments, consider the following approaches:

Ok, can you give a code example of how to do this?

**Comment by icyJoseph** [comment]
Isn't the issue here, that, browsers don't send fragments to the server? https://stackoverflow.com/a/940923

So, the rewrite never ever sees the fragment... You need to use a query param at best, to try and patch things together yourself.

**Comment by kouliavtsev** [comment]
@icyJoseph Are we talking about the same fragments? The `Text Fragments` I am referring to is relatively [new browser API](https://developer.mozilla.org/en-US/docs/Web/URI/Fragment/Text_fragments). The article you are pointing to is about 15 years old.

Or did I miss something here?

**Reply by icyJoseph** [reply]
Anything that comes after `#` in the URL is not sent by the browser. 

Try it yourself, go here:

- https://public.requestbin.com/r
- get the URL to the bin
- on your browser visit that bin, with a text fragment (and maybe a query to differentiate requests)
- see the logs back at the request bin site

**Reply by icyJoseph** [reply]
Did I convince you about the issue? 😄

**Reply by icyJoseph** [reply]
Also, btw, opening the link https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#:~:text=human with JavaScript enabled, blanks the fragment too 😄 

https://github.com/user-attachments/assets/d68eef6d-5d42-4d7b-a13e-9d729352b680

Though it does scroll. I guess MDN's SPA router is messing up with the fragment.

So, here is a Wikipedia link, https://en.wikipedia.org/wiki/React_(JavaScript_library)#:~:text=React%20can%20be%20used%20to, and if we inspect the Request Headers, we see what the browser sends to the server:

<img width="1403" alt="Screenshot 2024-11-07 at 20 59 52" src="https://github.com/user-attachments/assets/501ccac1-4132-4a7a-9490-804c7f5ac3a1">

So yeah, the browser's not sharing that with the server it communicates to.

**Reply by tombruijn** [reply]
I don't think the question is to receive the text fragment from the browser. We don't need it on the sever side.

In the Next.js frontend, the app is rewriting the URL from `/path#:~:text=for` to `/path` automatically. It doesn't do that for anchors, whether they exist or not. It only removes the text fragments from the browser's address bar.

**Reply by icyJoseph** [reply]
Aha, I thought OP was doing page rewrites 🤔 

So, then it is the same issue as MDN, that somehow rewrites the URL, dropping the text fragment? 

Does that mean that if I boot up any next.js app and load it with you these new fragments, I can expect to reproduce this?

**Reply by tombruijn** [reply]
> Does that mean that if I boot up any next.js app and load it with you these new fragments, I can expect to reproduce this?

I defer that back to @kouliavtsev. I don't use it enough to say what's normal behavior. If you know of a basic Next.js example app that's already live somewhere we can also test it against that one.

**Reply by icyJoseph** [reply]
No, I think I've got it ~ the OP MDN links don't necessarily work as they are presented, but I have managed to do some tests, and gain the context not properly transmitted. 

I'll do some debugging when I get time.

**Reply by icyJoseph** [reply]
Alright, so, this is a bit nasty...

`window.location.href`, includes good ol' document fragments, but it doesn't include text fragments...

A regular fragment:

<img width="1403" alt="Screenshot 2024-11-08 at 12 49 56" src="https://github.com/user-attachments/assets/14eb7ce3-acb5-416d-9b0b-4fe9bd8ff335">

vs...

<img width="1403" alt="Screenshot 2024-11-08 at 12 49 23" src="https://github.com/user-attachments/assets/424eebde-ff48-496b-85f9-07a122dbfbca">

🤷 

The Next.js router, as far as I can tell, uses `window.location.href` for a couple of things... but the href doesn't have it, neither do we get the fragment on the `hash` of `window.location`

<img width="787" alt="Screenshot 2024-11-08 at 12 52 19" src="https://github.com/user-attachments/assets/9923038b-1a16-4855-a12f-a86e92130b70">

> [!NOTE]  
> The second `window.location.href`, is run in a page with a url with a text fragment

The bottom line is, how do we even gain access in JS, to the text fragment? https://stackoverflow.com/a/74548006

**Reply by icyJoseph** [reply]
Update, on getting the text fragment... with a hack...

Source: https://web.dev/articles/text-fragments#obtaining_text_fragments_for_analytics_purposes

```js
new URL(performance.getEntries().find(({ type }) => type === 'navigate').name).hash;
```

<img width="1311" alt="Screenshot 2024-11-08 at 12 58 10" src="https://github.com/user-attachments/assets/4ce59564-4453-42a9-a670-dd043db1a33a">

🤔 what a funky feature...

**Reply by icyJoseph** [reply]
And, to wrap my answers here, https://github.com/WICG/scroll-to-text-fragment/issues/128

Which indicates that, in current implementations of this feature:

> For compatibility reasons, we strip the fragment directive from the URL before page script has a chance to run. Even if it was available to the page, it would be difficult for the page to know in all cases which piece of text the fragment matches, if any, since it would require effectively re-implementing the browser's implementation and running it on the page.
> 
> As mentioned in https://github.com/mozilla/standards-positions/issues/194#issuecomment-568581041-permalink:~:text=The%20marginalia%20use%20case%20described%20at%20https%3A%2F%2Findieweb.org%2Fmarginalia%20is%20not%20supported%20as%20well%20as%20it%20could%20be%2C%20since%20the%20fragment%20specifier%20is%20stripped%20from%20the%20URL.%20That%20significantly%20limits%20the%20ability%20of%20sites%20to%20show%20appropriate%20marginalia%20UX%20when%20the%20user%20loads%20a%20link%20to%20a%20specific%20text%20fragment.%20It%20could%20probably%20be%20addressed%20by%20exposing%20the%20fragment%20specifier%20via%20the%20FragmentDirective%2C%20modulo%20whatever%20the%20security%20considerations%20are%20with%20exposing%20this%20information%20to%20the%20page.%20As%20far%20as%20we%20can%20tell%2C%20those%20security%20considerations%20seem%20to%20be%20entirely%20driven%20by%20the%20search%20engine%20use%20case%2C%20and%20impose%20unnecessary%20constraints%20on%20the%20other%20use%20cases.%20That%20said%2C%20exposing%20the%20fragment%20specifier%20could%20be%20added%20in%20a%20V2%20of%20the%20API%3B%20perhaps%20along%20with%20allowing%20URLs%20to%20opt%20out%20of%20having%20it%20exposed.%20That%20would%20need%20to%20be,planned%20for%20in%20V1%20syntax%2C%20possibly.,-The of this feature, there exist interesting use cases for knowing the text fragment, such as [marginalia](https://indieweb.org/marginalia). We should expose this information so authors can make use of it for such cases.

So, it looks like, it is a known issue, that scripts can't really access the text fragment, besides looping through the performance entries 🤔 - which doesn't seem to be ideal/correct

**Reply by kouliavtsev** [reply]
@icyJoseph Great findings! Thank you for looking into this. It is what it is...

**Reply by icyJoseph** [reply]
Yeah, I remember when this was first proposed, people didn't really like it.

Sadly, it wasn't even that the browser doesn't send it to the server, but, for now, it also tries to hide it from the location API 😕 the performance access is questionable ~ only available as a means to get tracking in place
